arraycontext/impl/pytato/utils.py

```diff
@@ -0,0 +1,117 @@
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+from typing import Any, Dict, Set, Tuple, Mapping
+from pytato.array import Placeholder as BasePlaceholder
+from pytato.array import SizeParam as BaseSizeParam
+from pytato.array import ShapeType, TagsType, Array, DataWrapper
+from pytato.transform import CopyMapper
+from pytools import UniqueNameGenerator
+
+
+class Placeholder(BasePlaceholder):
+    """
+    An equality-comparable :class:`pytato.Placeholder`.
+    """
+    def __init__(self, name: str,
+                 shape: ShapeType,
+                 dtype: np.dtype,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, shape, dtype, tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
+
+
+class SizeParam(BaseSizeParam):
+    """
+    An equality-comparable :class:`pytato.SizeParam`.
+    """
+    def __init__(self,
+                 name: str,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, tags=tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
+
+
+class Normalizer(CopyMapper):
+    def __init__(self) -> None:
+        super().__init__()
+        self.bound_arguments: Dict[str, Any] = {}
+        self.vng = UniqueNameGenerator()
+        self.seen_inputs: Set[str] = set()
+
+    def map_data_wrapper(self, expr: DataWrapper) -> Array:
+        if expr.name is not None:
+            if expr.name in self.seen_inputs:
+                raise ValueError("Got multiple inputs with the name"
+                                 f"{expr.name} => Illegal.")
+            self.seen_inputs.add(expr.name)
+
+        name = self.vng("_actx_dw")
+        self.bound_arguments[name] = expr.data
+        return Placeholder(name=name,
+                           shape=tuple(self.rec(s) if isinstance(s, Array) else s
+                                       for s in expr.shape),
+                           dtype=expr.dtype,
+                           tags=expr.tags)
+
+    def map_size_param(self, expr: BaseSizeParam) -> Array:
+        if expr.name in self.seen_inputs:
+            raise ValueError("Got multiple inputs with the name"
+                             f"{expr.name} => Illegal.")
+        assert isinstance(expr.name, str)
```

Comment:
Out of curiosity, could this ever be not a string?

Reply:
This actually got me into thinking that handling `SizeParam` would be more than what we need to handle. raised a `NotImplementedError` for these for now.